### PR TITLE
Replace locks with Interlocked.Increment

### PR DESCRIPTION
**Pull request by xtqqczze**
# PR Summary

* Remove locks and replace with `Interlocked.Increment` and `Interlocked.Decrement` where possible.

Note: if overflow/underflow occurs, an exception will not be thrown, even if compiled in checked mode. `Debug.Assert` checks reverted as there is currently no checked build.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [NA] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [NA] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [NA] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [NA] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [NA] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by xtqqczze** [comment]
Documentation for [Interlocked.Increment](https://docs.microsoft.com/dotnet/api/system.threading.interlocked.increment) states: 

> This method handles an overflow condition by wrapping: if location = Int32.MaxValue, location + 1 = Int32.MinValue. No exception is thrown.

**Comment by xtqqczze** [comment]
> Documentation for [Interlocked.Increment(Int32)](https://docs.microsoft.com/dotnet/api/system.threading.interlocked.increment) states:
> 
> > This method handles an overflow condition by wrapping: if location = Int32.MaxValue, location + 1 = Int32.MinValue. No exception is thrown.

This is the same behaviour for the regular increment operator `++`, unless compiled in checked mode. So I will add some Debug.Asserts.

**Comment by xtqqczze** [comment]
> This is the same behaviour for the regular increment operator `++`, unless compiled in checked mode. So I will add some Debug.Asserts.

It seems PowerShell debug build does not check for arithmetic overflow. So `Debug.Assert` is unnecessary as there is no change in behaviour.

**Comment by xtqqczze** [comment]
rebased

File: src/Microsoft.Management.Infrastructure.CimCmdlets/CimAsyncOperation.cs [hunk 1]
```diff
@@ -107,13 +104,9 @@ protected void OperationDeletedHandler(object cimSession, OperationEventArgs act
...
 
-            lock (this.a_lock)
+            if (Interlocked.Decrement(ref this.operationCount) == 0)
             {
-                this.operationCount--;
-                if (this.operationCount == 0)
-                {
-                    this.moreActionEvent.Set();
-                }
+                this.moreActionEvent.Set();
```
Comment (iSazonov) [inline src/Microsoft.Management.Infrastructure.CimCmdlets/CimAsyncOperation.cs:109]: This was under lock before.

Comment (xtqqczze) [reply src/Microsoft.Management.Infrastructure.CimCmdlets/CimAsyncOperation.cs:109]: I see that other `Set`, `Wait`, `Reset` calls on the `moreActionEvent` object are not protected by a lock, therefore I assume the original intent was to protect only the `operationCount` field.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/WebServiceProxy.cs [hunk 2]
```diff
@@ -296,11 +296,7 @@ private string GenerateRandomName()
                 }
             }
 
-            string sequenceString;
-            lock (s_sequenceNumberLock)
-            {
-                sequenceString = (s_sequenceNumber++).ToString(CultureInfo.InvariantCulture);
-            }
+            string sequenceString = Interlocked.Increment(ref s_sequenceNumber).ToString(CultureInfo.InvariantCulture);
```
[the remark this answers is not in the document]
Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/WebServiceProxy.cs:299]: `s_sequenceNumber` is `UInt64` which is a readonly struct therefore `ToString()` should not need protecting.

`s_sequenceNumber` is referenced once (here) after ctor.

**Review by daxian-dbw — DISMISSED** [review]
LGTM

File: src/Microsoft.PowerShell.Commands.Management/commands/management/WebServiceProxy.cs [hunk 3]
```diff
@@ -296,11 +296,7 @@ private string GenerateRandomName()
                 }
             }
 
-            string sequenceString;
-            lock (s_sequenceNumberLock)
-            {
-                sequenceString = (s_sequenceNumber++).ToString(CultureInfo.InvariantCulture);
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/WebServiceProxy.cs:302]: `(s_sequenceNumber++).ToString(CultureInfo.InvariantCulture)` actually calls `ToString` on the original value of `s_sequenceNumber`, and then increment `s_sequenceNumber`. For example, running the following c# code:
```c#
	uint a = 1;
	string c = (a++).ToString();
	Console.WriteLine(c);
	Console.WriteLine(a);
```
you will get
```console
1
2
```
However, after your change, the incremental will happen before `ToString`, and that changes the original behavior.

File: src/Microsoft.PowerShell.ScheduledJob/ScheduledJobDefinition.cs [hunk 4]
```diff
@@ -1883,10 +1883,7 @@ public void Dispose()
 
         private static int GetCurrentId()
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.ScheduledJob/ScheduledJobDefinition.cs:1884]: We don't compile the scheduledjob code in GitHub repo, so please revert the changes here.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/WebServiceProxy.cs [hunk 5]
```diff
@@ -273,7 +274,6 @@ protected override void BeginProcessing()
         #region private
 
         private static ulong s_sequenceNumber = 1;
-        private static object s_sequenceNumberLock = new object();
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/WebServiceProxy.cs:276]: Actually, I just found that we don't even compile the `WebServiceProxy.cs` (see `Microsoft.PowerShell.Commands.Management.csproj` for the `<compile remove=` entries). So, please revert the changes in this file.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/WMIHelper.cs [hunk 6]
```diff
@@ -1532,10 +1532,8 @@ private void HandleChildJobStateChanged(object sender, JobStateEventArgs e)
             if (e.JobStateInfo.State == JobState.Blocked)
             {
                 // increment count of blocked child jobs
-                lock (_syncObject)
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/WMIHelper.cs:1535]: I just found that we don't even compile the `WMIHelper.cs` (see `Microsoft.PowerShell.Commands.Management.csproj` for the `<compile remove=` entries). So, please revert changes in this file.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

**Comment by daxian-dbw** [comment]
@iSazonov Please don't open this PR. It's been marked as "Waiting on Author" for 25 days, and the author didn't get to it. The author can re-open when ready to work on it again.
